bpo-38347: find Python scripts whose name contain a '-'

rpluem opened the pull request:
Allow Python scripts to contain a '-' in their filename when working
recursively.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-38347](https://bugs.python.org/issue38347) -->
https://bugs.python.org/issue38347
<!-- /issue-number -->

rpluem committed:
bpo-38347: find Python scripts whose name contain a '-'

Allow Python scripts to contain a '-' in their filename when working
recursively.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner reviewed (commented):
Would you mind to add a test to Lib/test/test_tools/test_pathfix.py?

rpluem committed:
bpo-38347: assume all files that end on '.py' are Python scripts

Assume all files that end on '.py' are Python scripts when working
recursively.

rpluem committed:
bpo-38347: add a test case for the recursive search

Add a test case for the recursive search including filenames with
a '-'.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

rpluem committed:
bpo-38347: use os.sep instead of the Unix specific '/'

Fix a test failure on Windows by using os.sep instead of the Unix
specific '/' as path separator.

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -14,16 +14,24 @@ class TestPathfixFunctional(unittest.TestCase):
-        self.addCleanup(support.unlink, support.TESTFN)
+        self.temp_directory = support.TESTFN + '.d'
+        self.addCleanup(support.rmtree, self.temp_directory)
+        os.mkdir(self.temp_directory)
```
Please move this code into test_recursive(). Other tests don't need that.

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -14,16 +14,24 @@ class TestPathfixFunctional(unittest.TestCase):
+        self.addCleanup(support.unlink, self.temp_file)
+
+    def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr='',
+                target_file=''):
```
I suggest to rename target_file to filename.

rpluem replied:
Fixed in 07f7158

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -44,55 +52,66 @@ def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr=''):
+               callable(method):
+                method(target_file=self.temp_directory, stderr=expected_stderr)
+
+    def test_pathfix(self, **kwargs):
```
I don't understand the purpose of the new **kwargs argument.

vstinner commented on Misc/NEWS.d/next/Tests/2019-10-02-12-27-58.bpo-38347.zdkCBX.rst:
```diff
@@ -0,0 +1 @@
+Add a test case for the recursive search including filenames with a '-'.
```
Please remove this NEWS entry. The other one is enough.

rpluem replied:
Done in 14f5aa0d351fddae1e798b5838f498c648bbcbda

rpluem replied:
Just as a comment: This causes one of the automatic checks to fail again, but I guess this is not a real blocker.

vstinner commented on Misc/NEWS.d/next/Tools/Demos/2019-10-02-09-48-42.bpo-38347.2Tq5D1.rst:
```diff
@@ -0,0 +1 @@
+Allow Python scripts to contain a '-' in their filename when working recursively.
```
You should mention pathfix somewhere. And this NEWS entry is outdated: it's no longer about "-", but any character.

rpluem replied:
Done in a68ad4207100b365d00ffbb745592408d1d55a98 and 05c39a5dd793c45fb6c2f2cff62b947a9576b352

rpluem committed:
bpo-38347: rename variable target_file to filename

No functional change.

rpluem committed:
bpo-38347: remove unneeded NEWS entry

rpluem committed:
bpo-38347: adjust comment to refelct recent changes

rpluem committed:
bpo-38347: forgot to mention pathfix.py

vstinner reviewed (changes requested):
I see that you updated your PR, but setUp() still creates a temporary directory (I asked to move that into the test_recursive() method), there is still "**kwargs", and now test_recursive() inspects self to call other functions... I don't understand what you are trying to do.

My expectation is that test_recursive() creates a temporary directory which contains one ".py" file which has a shebang, you can pathfix on the directory, and then you check that the .py file shebang is updated.

Please don't try to magically make other test methods work on a temporary directory. I prefer simple tests.

rpluem committed:
Revert "bpo-38347: rename variable target_file to filename"

This reverts commit 07f7158cfd3b6aa5e403317d7e7a3c959222ff5e.

rpluem committed:
Revert "bpo-38347: use os.sep instead of the Unix specific '/'"

This reverts commit 2dcdef4b36b48e85c3c53f28915c07ad40a8c4a5.

rpluem committed:
Revert "bpo-38347: add a test case for the recursive search"

This reverts commit 9c512f4c571df7cce922543473bbf39d2bfa5052.

rpluem committed:
bpo-38347: add a test case for the recursive search

Add a test case for the recursive search including filenames with
a '-'.

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -17,13 +17,17 @@ def setUp(self):
+        if filename == '':
+            filename = self.temp_file
+
         with open(self.temp_file, 'w', encoding='utf8') as f:
```
You should use filename here and below.

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -44,6 +48,22 @@ def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr=''):
+        self.temp_file = self.temp_directory + os.sep + \
+                         os.path.basename(support.TESTFN) + '-t.py'
+        temp_directory_basename = os.path.basename(self.temp_directory)
+        expected_stderr = f'recursedown(\'{temp_directory_basename}\')\n'
```
I suggest:

```
tmpdir = support.TESTFN
self.addCleanup(support.rmtree, tmpdir)
os.mkdir(tmpdir)
expected_stderr = f"recursedown('{os.path.basename(tmpdir)}')\n"
... self.pathfix(..., directory=tmpdir, ...) ...
```

By the way, the self.temp_file attribute is useless. You can simply remove the setUp() method. If you replace pathfix() `filename` parameter with a new `directory` parameter, you can write something like:

```
if directory:
  pathfix_arg = directory
  filename = os.path.join(directory, 'script.py')
else:
  filename = support.TESTFN
  self.addCleanup(support.unlink, filename) 
  pathfix_arg = filename

... subprocess.run([sys.executable, self.script, *pathfix_flags, '-n', pathfix_arg], ...) ...
```

rpluem replied:
Wouldn't adding the self.addCleanup call to pathfix cause the cleanup to be added multiple times (each time we call pathfix)? And if yes, is this really desired?
When are the cleanups run? After each single test or after all tests of this class have been executed?

vstinner replied:
Right. One solution is to use try/finally to manually call support.unlink(filename) if directory is not set.

rpluem replied:
Any reason why you don't like the setUp method and the addCleanup approach and prefer a local try / finally solution?

vstinner replied:
Well, you can keep setUp() as it is if you prefer.

rpluem committed:
bpo-38347: Use local variables with better names

Use local variables with better names instead of class attributes.

vstinner commented on Misc/NEWS.d/next/Tools/Demos/2019-10-02-09-48-42.bpo-38347.2Tq5D1.rst:
```diff
@@ -0,0 +1 @@
+pathfix.py: Assume all files that end on '.py' are Python scripts when working recursively.
```
I'm not sure how you created this file, but blurb uses the directory "Misc/NEWS.d/next/Tools-Demos/", not "Misc/NEWS.d/next/Tools/Demos/".

rpluem replied:
Weird. It was created by blurb in d8e988efe85bfef307258d06bbeae72e4a21d1fe. Fixed in a17410299b88bb3063daaed18f67efe70c9f2a2c.

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -44,6 +51,19 @@ def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr=''):
+        tmpdir = support.TESTFN + '.d'
+        self.addCleanup(support.rmtree, tmpdir)
+        os.mkdir(tmpdir)
+        expected_stderr = f'recursedown(\'{os.path.basename(tmpdir)}\')\n'
```
```suggestion
        expected_stderr = f"recursedown('{os.path.basename(tmpdir)}')\n"
```

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -14,25 +14,32 @@ class TestPathfixFunctional(unittest.TestCase):
+    def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr='',
+                directory=''):
+        if directory:
+            filename = os.path.join(directory, 'script-A_1.py')
```
```suggestion
            filename = os.path.join(directory, 'script.py')
```

rpluem replied:
This is on purpose. With 'script.py' the test would not have detected the issue that this pull request fixes. And I think a test case for a fix should fail without the fix and succeed afterwards. We can discuss though how we add a capital letter, a '_', a number and a '-' more nicely to the name. I am also fine if I should add a comment that explains the weird script name.

vstinner replied:
Yeah, please add a comment like "# bpo-38347: Test filename containing uppercase, "-", "_" and digit". Otherwise, someone may remove it because it looks like a typo.

rpluem replied:
Done in 56eaa9441a3f56420817634d910c2575ff288131.

rpluem committed:
bpo-38347: move NEWS to the correct location

rpluem committed:
bpo-38347: Use double quotes instead of single ones

Co-Authored-By: Victor Stinner <vstinner@python.org>

rpluem committed:
bpo-38347: Add comment to explain the name of the script

vstinner commented on Lib/test/test_tools/test_pathfix.py:
```diff
@@ -14,25 +14,34 @@ class TestPathfixFunctional(unittest.TestCase):
+    def pathfix(self, shebang, pathfix_flags, exitcode=0, stdout='', stderr='',
+                directory=''):
+        if directory:
+            # bpo-38347: Test filename should contain lowercase, upercase,
```
type: uppercase
```suggestion
            # bpo-38347: Test filename should contain lowercase, uppercase,
```

vstinner reviewed (approved):
LGTM, but please fix the typo. It seems like I'm not allowed to push into your branch.

rpluem committed:
Fix typo

Co-Authored-By: Victor Stinner <vstinner@python.org>
